Migrate to Numpy v2.0

Maintenance for moving to numpy v2.0 (need issue links) ([migration guide](https://numpy.org/doc/stable/numpy_2_0_migration_guide.html))

- [ ] Need prototype(s) for how extensive the changes might be
- [ ] Will be done on a long-running branch to integrate the changes. `main` will be auto-merged into the branch
- [x] numpy v2 supports python 3.9-3.12 ([python supported versions](https://devguide.python.org/versions/))
- [x] boost v1.84 supports newer numpy [boost::python docs](https://www.boost.org/doc/libs/1_84_0/libs/python/doc/html/numpy/index.html) https://github.com/mantidproject/mantid/pull/37766
- [x] Move to matplotlib >=3.8.4 https://github.com/mantidproject/mantid/pull/38077/

Mantid will not move to matplotlib=3.9 until later because it requires qt=6

> I believe quasielasticbayes (v0.2.2) should now work with Numpy 2 https://anaconda.org/conda-forge/quasielasticbayes

Thanks @robertapplin !

Tested with:
- python 3.10.14
- numpy 2.1.1
- matplotlib 3.8.4
- quasielasticbayes 0.2.2

=> Mantid workbench launches
=> 
```
FAIL: test_QSe_Run (BayesQuasiTest.BayesQuasiTest)
Test Stretched Exponential fit for BayesQuasi
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/perenon/_Dev/mantid/mantill/Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/BayesQuasiTest.py", line 79, in test_QSe_Run
    self._validate_QSe_value(result, fit_group)
  File "/home/perenon/_Dev/mantid/mantill/Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/BayesQuasiTest.py", line 243, in _validate_QSe_value
    self.assertAlmostEqual(result.dataY(0)[0], 81.12644, delta=1e-4)
AssertionError: np.float64(81.07385699682999) != 81.12644 within 0.0001 delta (np.float64(0.05258300317001385) difference)

```
That will need further investigations. I will keep you up to date !

--- a/Framework/PythonInterface/mantid/plots/datafunctions.py
+++ b/Framework/PythonInterface/mantid/plots/datafunctions.py
@@ -398,7 +398,7 @@ def get_bin_indices(workspace):
     else:
         # the following two lines can be replaced by np.isin when > version 1.7.0 is used on RHEL7
         total_range = np.asarray(range(total_range))
-        indices = np.where(np.in1d(total_range, monitors_indices, invert=True).reshape(total_range.shape))
+        indices = np.where(np.isin(total_range, monitors_indices, invert=True).reshape(total_range.shape))
         # this check is necessary as numpy may return a tuple or a plain array based on platform.
         indices = indices[0] if isinstance(indices, tuple) else indices
         return indices

--- a/Framework/PythonInterface/mantid/plots/mantidaxes.py
+++ b/Framework/PythonInterface/mantid/plots/mantidaxes.py
@@ -1526,7 +1526,8 @@ class MantidAxes3D(Axes3D):
 
             # This is a bit of a hack, should be able to remove
             # when matplotlib supports plotting masked arrays
-            poly_c._A = safe_masked_invalid(poly_c._A)
+            if poly_c._A is not None:
+                poly_c._A = safe_masked_invalid(poly_c._A)
 
         # Create a copy of the original data points because data are set to nan when the axis limits are changed.
         self.original_data_surface = copy.deepcopy(poly_c._vec)

--- a/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/FindPeaksAutomatic.py
+++ b/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/FindPeaksAutomatic.py
@@ -44,7 +44,7 @@ class FindPeaksAutomatic(DataProcessorAlgorithm):
         # Input parameters
         self.declareProperty("SpectrumNumber", 1, doc="Spectrum number to use", validator=IntBoundedValidator(lower=0))
         self.declareProperty("StartXValue", 0.0, doc="Value of X to start the search from")
-        self.declareProperty("EndXValue", np.Inf, doc="Value of X to stop the search to")
+        self.declareProperty("EndXValue", np.inf, doc="Value of X to stop the search to")
         self.declareProperty(
             "AcceptanceThreshold",
             0.01,

--- a/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/FitGaussianPeaks.py
+++ b/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/FitGaussianPeaks.py
@@ -261,7 +261,7 @@ class FitGaussianPeaks(DataProcessorAlgorithm):
         y_fit = y_fit[np.nonzero(y_data)]
         y_data = y_data[np.nonzero(y_data)]
         if len(y_fit) < 1:
-            return -np.Inf
+            return -np.inf
         y_log = np.log(np.abs(y_fit))
         return np.sum(-y_fit + y_data * y_log)
 

--- a/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLSumForeground.py
+++ b/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLSumForeground.py
@@ -285,9 +285,9 @@ class ReflectometryILLSumForeground(DataProcessorAlgorithm):
         if dist != 0.0:
             det_point_1 = ws.spectrumInfo().position(0)
             det_point_2 = ws.spectrumInfo().position(20)
-            beta = numpy.math.atan2((det_point_2[0] - det_point_1[0]), (det_point_2[2] - det_point_1[2]))
-            x_vs_y = numpy.math.sin(beta) * dist
-            mz = numpy.math.cos(beta) * dist
+            beta = numpy.atan2((det_point_2[0] - det_point_1[0]), (det_point_2[2] - det_point_1[2]))
+            x_vs_y = numpy.sin(beta) * dist
+            mz = numpy.cos(beta) * dist
             if instr == "D17":
                 mx = x_vs_y
                 my = 0.0

--- a/Framework/PythonInterface/test/python/mantid/utils/nomad/diagnosticsTest.py
+++ b/Framework/PythonInterface/test/python/mantid/utils/nomad/diagnosticsTest.py
@@ -434,12 +434,12 @@ class NOMADMedianDetectorTestTest(unittest.TestCase):
         levels = tester.tube_collevel  # shape = (TUBE_COUNT,)
         assert len(levels) == tester.TUBE_COUNT
         first1, last1 = tester.TUBES_IN_EIGHTPACK * 77, tester.TUBES_IN_EIGHTPACK * 78
-        assert np.all(levels[:first1]) == CollimationLevel.Empty
-        assert np.all(levels[first1:last1]) == CollimationLevel.Half
+        assert np.all(levels[:first1] == CollimationLevel.Empty)
+        assert np.all(levels[first1:last1] == CollimationLevel.Half)
         first2, last2 = tester.TUBES_IN_EIGHTPACK * 95, tester.TUBES_IN_EIGHTPACK * 96
-        assert np.all(levels[last1:first2]) == CollimationLevel.Empty
-        assert np.all(levels[first2:last2]) == CollimationLevel.Full
-        assert np.all(levels[last2:]) == CollimationLevel.Empty
+        assert np.all(levels[last1:first2] == CollimationLevel.Empty)
+        assert np.all(levels[first2:last2] == CollimationLevel.Full)
+        assert np.all(levels[last2:] == CollimationLevel.Empty)
 
     def test_panel_median(self):
         tester = self.tester1

--- a/Framework/PythonInterface/test/python/plugins/algorithms/SaveP2DTest.py
+++ b/Framework/PythonInterface/test/python/plugins/algorithms/SaveP2DTest.py
@@ -33,7 +33,7 @@ class SaveP2DTest(unittest.TestCase):
         """Create a dummy workspace for testing purposes"""
         xData = [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]  # d values for one spectrum (one dPerpendicular value)
         yData = ["1", "2", "3"]  # dPerpendicular binedges
-        zData = [1.0, 2.0, 3.0, -1.0, 0.0, np.NaN, 3.0, 1.0, 4.0]  # intensity values
+        zData = [1.0, 2.0, 3.0, -1.0, 0.0, np.nan, 3.0, 1.0, 4.0]  # intensity values
         eData = [0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1]  # error values
 
         # used to join all spectra

--- a/Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/FitGaussianPeaksTest.py
+++ b/Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/FitGaussianPeaksTest.py
@@ -172,7 +172,7 @@ class FitGaussianPeaksTest(unittest.TestCase):
 
     def test_parse_fit_table_marks_peaks_for_refitting_if_error_larger_than_value(self):
         peaks = [(35.2, 0.4), (25.03, 0.1), (10.03, 0.05)]
-        peaks += [(20.003, 40.22), (75.15, 0.2), (5.2, np.NaN)]
+        peaks += [(20.003, 40.22), (75.15, 0.2), (5.2, np.nan)]
         fit_table = self.simulate_fit_parameter_output(peaks, 100.034)
         data_table = CreateEmptyTableWorkspace()
 

--- a/buildconfig/CMake/PyStoG.cmake
+++ b/buildconfig/CMake/PyStoG.cmake
@@ -1,8 +1,8 @@
 include(ExternalProject)
 
-set(_PyStoG_VERSION f184ebf9a72aae48ae0d5267fd6ab2e7df0988f6) # v0.4.7
+set(_PyStoG_VERSION 7b5492d98817024f2b62867bc2a82fc23184b777) # v0.5.0
 set(_PyStoG_download_dir ${CMAKE_CURRENT_BINARY_DIR}/../PyStoG-download)
-set(_PyStoG_source_dir ${_PyStoG_download_dir}/src/PyStoG/pystog)
+set(_PyStoG_source_dir ${_PyStoG_download_dir}/src/PyStoG/src/pystog)
 set(_PyStoG_source_test_dir ${_PyStoG_download_dir}/src/PyStoG/tests)
 set(_PyStoG_scripts_dir ${CMAKE_CURRENT_BINARY_DIR}/pystog)
 set(_PyStoG_test_root_dir ${CMAKE_CURRENT_BINARY_DIR}/test/pystog)

--- a/conda/recipes/conda_build_config.yaml
+++ b/conda/recipes/conda_build_config.yaml
@@ -63,7 +63,7 @@ nexus:
   - 4.4.*
 # We follow conda-forge and build with the lowest supported version of numpy for forward compatibility.
 numpy:
-  - 1.24.*
+  - 2.0.*
 
 matplotlib:
   - 3.8.*

--- a/conda/recipes/mantid-developer/meta.yaml
+++ b/conda/recipes/mantid-developer/meta.yaml
@@ -28,7 +28,7 @@ requirements:
     - muparser {{ muparser }}
     - nexus {{ nexus }}
     - ninja {{ ninja }}
-    - numpy>=1.24,<1.27 # This is intentionally different to conda_build_config.yaml, will address later.
+    - numpy>=2.0,<2.2 # This is intentionally different to conda_build_config.yaml, will address later.
     - occt {{ occt }}
     - pip {{ pip }}
     - poco {{ poco }}

--- a/conda/recipes/mantid/meta.yaml
+++ b/conda/recipes/mantid/meta.yaml
@@ -63,7 +63,7 @@ requirements:
     - lib3mf  # [win]
     - mkl {{ mkl }} # [win]
     - nexus {{ nexus }}
-    - {{ pin_compatible("numpy", upper_bound="1.27") }}
+    - {{ pin_compatible("numpy", upper_bound="2.2") }}
     - occt {{ occt }}
     - pycifrw
     - python

--- a/docs/source/release/v6.12.0/Diffraction/Powder/New_features/38404.rst
+++ b/docs/source/release/v6.12.0/Diffraction/Powder/New_features/38404.rst
@@ -1,1 +1,1 @@
-- Move `pystog <https://github.com/neutrons/pystog>`_ from v0.2.7 to v0.4.7
+- Move `pystog <https://github.com/neutrons/pystog>`_ from v0.2.7 to v0.5.0 which supports numpy v2

--- a/docs/source/release/v6.12.0/Framework/Dependencies/New_features/38420.rst
+++ b/docs/source/release/v6.12.0/Framework/Dependencies/New_features/38420.rst
@@ -1,0 +1,1 @@
+- Drop support for NumPy version 1. We now build against NumPy v2.0 and support up to v2.1. `Read about the changes <https://numpy.org/news/#numpy-200-released>`_. **Users should note that NumPy 2 introduces some breaking API changes. See the `NumPy 2 Migration Guide <https://numpy.org/devdocs/numpy_2_0_migration_guide.html>`_ for more details**

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,7 +1,7 @@
 [tool.ruff]
 line-length = 140
 # https://beta.ruff.rs/docs/rules/
-lint.select = ["C90", "E", "F", "RUF100", "W"]
+lint.select = ["C90", "E", "F", "RUF100", "W", "NPY201"]
 lint.ignore = ["E722", # bare except
 ]
 exclude = [

--- a/qt/python/mantidqt/mantidqt/widgets/codeeditor/test/test_completion.py
+++ b/qt/python/mantidqt/mantidqt/widgets/codeeditor/test/test_completion.py
@@ -54,7 +54,9 @@ class CodeCompletionTest(unittest.TestCase):
         self._run_check_call_tip_generated("import numpy as np\n# My code", r"np\.asarray\(a, \[dtype\], .*\)")
 
     def test_numpy_call_tips_generated_handling_wildcards_properly_if_numpy_imported_in_script(self):
-        self._run_check_call_tip_generated("import numpy as np\n# My code", r"np\.asarray\(a, \[dtype\], \[order\], \*, \[like\]\)")
+        self._run_check_call_tip_generated(
+            "import numpy as np\n# My code", r"np\.asarray\(a, \[dtype\], \[order\], \*, \[device\], \[copy\], \[like\]\)"
+        )
 
     def test_call_tips_generated_if_syntax_errors_in_script(self):
         self._run_check_call_tip_generated("from mantid.simpleapi import *\n print 'Hello', 'World'", "Rebin")

--- a/qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/ElementalAnalysis/PeriodicTable/PeakSelector/peak_selector_view.py
+++ b/qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/ElementalAnalysis/PeriodicTable/PeakSelector/peak_selector_view.py
@@ -34,19 +34,19 @@ def valid_data(peak_data):
         return False
     if peak_data["Primary"] is not None:
         for x_pos in peak_data["Primary"].values():
-            if not value_in_bounds(x_pos, 0.0, np.Inf):
+            if not value_in_bounds(x_pos, 0.0, np.inf):
                 return False
     if peak_data["Secondary"] is not None:
         for x_pos in peak_data["Secondary"].values():
-            if not value_in_bounds(x_pos, 0.0, np.Inf):
+            if not value_in_bounds(x_pos, 0.0, np.inf):
                 return False
     if "Gammas" in data_label and peak_data["Gammas"] is not None:
         for x_pos in peak_data["Gammas"].values():
-            if not value_in_bounds(x_pos, 0.0, np.Inf):
+            if not value_in_bounds(x_pos, 0.0, np.inf):
                 return False
     if "Electrons" in data_label and peak_data["Electrons"] is not None:
         for x_pos in peak_data["Electrons"].values():
-            if not value_in_bounds(x_pos, 0.0, np.Inf):
+            if not value_in_bounds(x_pos, 0.0, np.inf):
                 return False
 
     return True

--- a/scripts/Inelastic/CrystalField/normalisation.py
+++ b/scripts/Inelastic/CrystalField/normalisation.py
@@ -37,7 +37,7 @@ def _get_normalisation(nre, bnames):
     for bname in bnames:
         bdict = {bname: 1}
         ee, vv, ham = CFEnergy(nre, **bdict)
-        Omat = np.mat(ham)
+        Omat = np.asmatrix(ham)
         norm = np.trace(np.real(Omat * np.conj(Omat))) / (2 * J + 1)
         retval[bname] = np.sqrt(np.abs(norm)) * np.sign(norm)
     return retval

--- a/scripts/Inelastic/Direct/PropertiesDescriptors.py
+++ b/scripts/Inelastic/Direct/PropertiesDescriptors.py
@@ -1685,10 +1685,10 @@ class RotationAngle(PropDescriptor):
         # pylint: disable=protected-access
         offset = self._mot_offset._offset
         if offset is None:
-            return np.NaN
+            return np.nan
         log_val = self._read_ws_logs(workspace)
         if log_val is None:
-            return np.NaN
+            return np.nan
         else:
             return offset + log_val
 

--- a/scripts/Muon/MaxentTools/opus.py
+++ b/scripts/Muon/MaxentTools/opus.py
@@ -10,7 +10,7 @@ import numpy as np
 def OPUS(x, SAVETIME_i2, PULSESHAPE_convol, DETECT_a, DETECT_b, DETECT_e):
     npts = DETECT_e.shape[0]
     n = x.shape[0]
-    y = np.zeros([SAVETIME_i2], dtype=np.complex_)
+    y = np.zeros([SAVETIME_i2], dtype=np.complex128)
     y[:n] = x * PULSESHAPE_convol
     y2 = np.fft.ifft(y) * SAVETIME_i2  # SN=+1, inverse FFT without the 1/N
     ox = (np.outer(np.real(y2[:npts]), DETECT_a) + np.outer(np.imag(y2[:npts]), DETECT_b)) * DETECT_e[:, np.newaxis]

--- a/scripts/Muon/MaxentTools/tropus.py
+++ b/scripts/Muon/MaxentTools/tropus.py
@@ -10,7 +10,7 @@ import numpy as np
 def TROPUS(ox, SAVETIME_i2, PULSESHAPE_convol, DETECT_a, DETECT_b, DETECT_e):
     npts, ngroups = ox.shape
     n = PULSESHAPE_convol.shape[0]
-    y = np.zeros([SAVETIME_i2], dtype=np.complex_)
+    y = np.zeros([SAVETIME_i2], dtype=np.complex128)
     y[:npts] = np.dot(ox, DETECT_a) * DETECT_e + 1.0j * np.dot(ox, DETECT_b) * DETECT_e
     y2 = np.fft.fft(y)  # SN=-1 meaning forward fft, scale is OK
     x = np.real(y2)[:n] * np.real(PULSESHAPE_convol) + np.imag(y2)[:n] * np.imag(PULSESHAPE_convol)

--- a/scripts/Muon/MaxentTools/zft.py
+++ b/scripts/Muon/MaxentTools/zft.py
@@ -16,7 +16,7 @@ import numpy as np
 def ZFT(f, PULSESHAPE_convol, DETECT_e, SAVETIME_i2):
     n = f.shape[0]
     npts = DETECT_e.shape[0]
-    y = np.zeros([SAVETIME_i2], dtype=np.complex_)
+    y = np.zeros([SAVETIME_i2], dtype=np.complex128)
     y[:n] = f * PULSESHAPE_convol
     y2 = np.fft.ifft(y) * SAVETIME_i2  # SN=+1 meaning inverse FFT without the 1/N scale factor
     return np.real(y2[:npts]) * DETECT_e, np.imag(y2[:npts]) * DETECT_e

--- a/scripts/pychop/Instruments.py
+++ b/scripts/pychop/Instruments.py
@@ -1001,7 +1001,7 @@ class Instrument(object):
             try:
                 etrans = float(etrans)
             except TypeError:
-                etrans = np.asfarray(etrans)
+                etrans = np.asarray(etrans, dtype=float)
         res = obj.getResolution(etrans)
         if return_polynomial:
 

--- a/scripts/test/Abins/AbinsBroadeningTest.py
+++ b/scripts/test/Abins/AbinsBroadeningTest.py
@@ -140,7 +140,7 @@ class BroadeningTest(unittest.TestCase):
         # Normal scheme reproduces area as well as total;
         freq_points, full_spectrum = broadening.broaden_spectrum(frequencies, bins, s_dft, sigma, scheme="normal")
         self.assertAlmostEqual(
-            np.trapz(spectrum, x=freq_points),
+            np.trapezoid(spectrum, x=freq_points),
             pre_broadening_total * (bins[1] - bins[0]),
         )
         self.assertAlmostEqual(sum(spectrum), pre_broadening_total)

